### PR TITLE
Optimize queries with references used in generated

### PR DESCRIPTION
**Pull request by bogensberger**
columns

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import java.util.Collection;
+import java.util.Map;
+
+@Singleton
+public class GeneratedColumnComparisonReplacer {
+
+    private static final Predicate<Symbol> IS_GENERATED_COLUMN = new Predicate<Symbol>() {
+        @Override
+        public boolean apply(@Nullable Symbol input) {
+            assert input != null;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java]: no need for the assertion,  ``input instanceof Reference`` will return false if input is null

File: sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                return symbol;
+            } else {
+                Context ctx = new Context(referencedSingleReferences);
+                return process(symbol, ctx);
+            }
+        }
+
+        @Override
+        public Symbol visitFunction(Function function, Context context) {
+            if (Operators.COMPARISON_OPERATORS.contains(function.info().ident().name()) &&
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java]: first contains check is unnecessary

File: sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                    function.info().ident().name().equals(EqOperator.NAME)) {
+                Reference reference = null;
+                Symbol otherSide = null;
+                for (int i = 0; i < function.arguments().size(); i++) {
+                    Symbol arg = function.arguments().get(i);
+                    if (arg instanceof Field) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java]: in which cases to we have Fields here? resolving here shouldn't be necessary

File: sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                                                      Arrays.asList(genColReference.info().type(), wrapped.valueType())), DataTypes.BOOLEAN);
+                Function comparisonFunction = new Function(comparisonFunctionInfo, Arrays.asList(genColReference, wrapped));
+                return new Function(AndOperator.INFO, Arrays.<Symbol>asList(function, comparisonFunction));
+
+            }
+            return function;
+        }
+
+        private Symbol wrapInGenerationExpression(Symbol wrapMeLikeItsHot, Reference generatedReference) {
+            ReferenceReplacer replacer = new ReferenceReplacer();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java]: replacer can be a constant field

File: sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java (hunk 3 above)
Comment (bogensberger) [inline sql/src/main/java/io/crate/analyze/GeneratedColumnComparisonReplacer.java]: the function arguments are always fields

File: sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java [hunk 5]
```diff
@@ -55,7 +55,8 @@
     private static final EnumSet<Join.Type> ALLOWED_JOIN_TYPES = EnumSet.of(Join.Type.CROSS, Join.Type.INNER);
 
     @Inject
-    public RelationAnalyzer(AnalysisMetaData analysisMetaData) {
+    public RelationAnalyzer(AnalysisMetaData analysisMetaData,
+                            GeneratedColumnComparisonReplacer generatedColumnComparisonReplacer) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java]: unused?

**Comment by mfussenegger** [comment]
ack

**Pull request merged**